### ISSUE TITLE
Add Immersive Engineering Uranium (Minecraft 1.10)

### ISSUE THREAD

**Issue by LiaungYip** [reteo/Sprocket#14]
With reference to https://github.com/reteo/Sprocket/pull/6 and https://github.com/reteo/Sprocket/pull/13,

Immersive Engineering for Minecraft 1.10 has added a brand new `oreUranium` block (`immersiveengineering:ore:5`). The spawn parameters in `immersiveengineering.cfg` are as follows:

```
        # Generation config for Uranium Ore.
        # Parameters: Vein size, lowest possible Y, highest possible Y, veins per chunk, chance for vein to spawn (out of 100). Set vein size to 0 to disable the generation
        I:ore_uranium <
            4
            8
            24
            2
            60
         >
```

I'm comfortable with the mechanics of adding this to the `.ods` spreadsheet, then putting this through `sprocket-csv.py` and `sprocket.py` to generate XML config files for COG.

However, I'm not sure **what exact numeric values to put in!**

Is there any guidance or methodology on how to translate from 'vanilla style' cluster generation parameters (here, 4 ores per vein, 2 * 60% veins per chunk) to COG parameters like "size of motherlode"?

Sidenote: In an 80x80 patch of the world, there were 62 uranium ore blocks, as counted by WorldEdit `//replace immersiveengineering:ore:5 glass`.

**Comment by reteo** [comment]
Most of the time, I went through the source code to the mods themselves, looking for anything associated with WorldGenMineable; sometimes the numbers would be right there, other times, they would be associated with variables that class pulls in.  If the source code wasn't available on sites like Github, I would often decompile them using a bytecode decompiler/viewer.  There's no other way to get an accurate estimate, because there's always the chance that veins should have spawned where empty air is located.

**Comment by LiaungYip** [comment]
Thanks for the information.

To clarify, I'm comfortable with defining the parameters for a Vanilla/StandardGen distribution. That's ores per vein, veins per chunk, and upper/lower Y values, as you've described.

The thing I'm unsure of is how to define a LayeredVeins distribution that is faithful to the original distribution of ore. Is there any method to working out variables like `Motherlode Size` and `Branch length`, or is it done by trial and error?

**Comment by LiaungYip** [comment]
My apologies - the formulas to work out the parameters are all right there in the .ODS spreadsheet.

I got confused because only a few rows were showing in Microsoft Excel and thought the spreadsheet could only be opened in LibreOffice. Actually, the majority of rows were hidden by table filtering.

**Comment by reteo** [comment]
Yeah, my apologies.  I sometimes hide columns to double-check my work from time to time.  I must have forgotten to undo the process before committing.

**Comment by LiaungYip** [comment]
All good. Actually, I've just finished converting that spreadsheet over to Microsoft Excel format (it still works in LibreOffice 5).

That's at https://github.com/LiaungYip/Sprocket/raw/xlsx-not-ods/COG%20Ore%20Settings%20Chart.xlsx .

Advantage of having it in Microsoft Excel format is that I can now use Python libraries (i.e. `xlrd`) to read the Microsoft Excel file *directly*, which saves the manual, error prone steps of converting the ODS file to CSV. (There was a bug/feature where the first 19 rows of input were skipped by `sprocket-csv.py` - which is meant to skip the 19 rows of junk at the top of the spreadsheet. But when `lawremi` saved the CSV, he manually trimmed those 19 rows off already!)

It also means there's only one source of the truth (an xlsx file) rather than two sources of truth (an ODS file and a CSV file, as is the case currently in `lawremi`'s fork.) That should eliminate one source of confusion for contributors.

I also took the opportunity to make all the formulas consistent. I found a few errors/quirks in the original calculations while doing so. For example, the Thaumcraft4 ores were using ore density parameters for Stratum Clouds instead of Strategic Clouds.

Net result is that it should be substantially easier for future maintainers to add new ores, and keep the thing up to date.

**Comment by LiaungYip** [comment]
Also added some instructions to the spreadsheet so people know what to do with it...

![image](https://user-images.githubusercontent.com/2173684/27003791-9bf867dc-4e40-11e7-950b-30bb2d699490.png)

**Comment by reteo** [comment]
The main reason I used LibreOffice was because I use Linux.  Obviously, using Office would shut out those people, like myself, who want to avoid Windows on their PCs.

**Comment by reteo** [comment]
First, those 14 rows were important, as they contained the calculations to align the spreadsheet values with the default values in COG's main configuration XML.  

Next, the disadvantage of using Excel is that it causes even more manual, error-prone steps for those people, like myself, who choose to use the Linux operating sustem on their computers.  LibreOffice is open-source, so it is available in all platforms, unlike Microsoft's offering which is available in Windows first, and maaaaaybe Mac, if they can be bothered with it.

I'm sure there's a way to directly access ODF files from Python, in which case it can be a source of "truth."  I'm just not coding whiz enough to really know what it is, hence the easier step of just coding with CSV support.

As for Thaumcraft 4 ores; the reason I chose stratum was to keep them in a solid mass, rather than sprinkled about; I was under the impression that magic power would, itself, be in a solid area, affecting everything in that area.

And, finally, as I previously mentioned, the net result would likely only be easier for those who use Windows.

**Comment by LiaungYip** [comment]
Thanks for the feedback!

> First, those 14 rows were important, as they contained the calculations to...

The calculations on the first 14 rows have been retained, just moved to a second worksheet.

![image](https://user-images.githubusercontent.com/2173684/27004362-88b693f0-4e4b-11e7-87d9-777b807f6767.png)

> Next, the disadvantage of using Excel is that it causes even more manual, error-prone steps for those people, like myself, who choose to use Linux ... LibreOffice is open-source, so it is available in all platforms...

I used to be an Arch Linux user (before the corporate world beat me into using Windows), so I understand and sympathise!

In this case, I am testing and making sure that the `.xlsx` file can **be edited in the latest version of LibreOffice** (LibreOffice Calc 5, at the time of writing) with no loss of functionality.

Note:
  * The LibreOffice file utterly broke for me when I tried to open it in Excel - all the formulas were destroyed. Instead of `$H75/$G75/2` I just had `0.876`, or whatever. That was really confusing.
  * But the Excel file seems to open fine in LibreOffice.
  * So as far as compatibility goes, seems it's better to keep it in Excel format.

> I'm sure there's a way to directly access ODF files from Python...

The libraries for dealing with ODF in Python [aren't very good](https://pypi.python.org/pypi/pyexcel-ods3/0.3.2).

The libraries for dealing with Excel files are easy to use, well documented, and battle-hardened through frequent use.

----

> As for Thaumcraft 4 ores; the reason I chose stratum...

Aaah. So the formulas were right, but the name of the distribution wasn't. I can fix that!

(Though it's sort of academic, seeing as Thaumcraft 4 doesn't exist on Minecraft 1.10+ yet.)

![image](https://user-images.githubusercontent.com/2173684/27004559-39c2eb6a-4e4e-11e7-90f9-12bd7a3f0d29.png)

----

In any case, I'll continue along this path and see if I can come to a solution that is easy to use, works on all operating systems, and is compatible with LibreOffice.

If it works, we all win. If it doesn't work, I'll bury it in the back yard, and no-one will ever know it existed. :)
